Do not change path info when loading a base version

mirefly opened the pull request:
Fixes #1313

mirefly commented on src/reducers/versions.tsx:
```diff
@@ -1253,8 +1259,23 @@ export const createReducer = ({
+        const { updatePathInfo, version } = action.payload;
         const internalVersion = createInternalVersion(version);
+
+        if (updatePathInfo === false) {
```
I intentionally did this `updatePathInfo===false`, so that undefined `updatePathInfo` could be treated as `true`. It seems to be a bad idea, so I made some changes to set the default value explicitly.

mirefly commented on src/reducers/versions.tsx:
```diff
@@ -244,6 +244,7 @@ export const actions = {
   loadVersionInfo: createAction('LOAD_VERSION_INFO', (resolve) => {
     return (payload: {
       version: ExternalVersionWithContent | ExternalVersionWithDiff;
+      updatePathInfo?: boolean;
```
`loadVersionInfo` was used a lot (60+ times) in tests. It seems to be annoying to pass `updatePathInfo: true` for all. I updated the patch to make the default value (`true`) of `updatePathInfo` more clear.

mirefly commented:
Thanks for reviewing this patch @bobsilverberg . I updated the patch. I tried to make the default value of `updatePathInfo` more explicit, but still optional.

bobsilverberg commented on src/reducers/versions.tsx:
```diff
@@ -242,9 +242,13 @@ export const actions = {
   loadVersionInfo: createAction('LOAD_VERSION_INFO', (resolve) => {
-    return (payload: {
+    return ({
+      updatePathInfo = true,
```
Thanks for the update @mirefly. I know I asked for the default to be more explicit, but I find this way of doing it to be inconsistent with the way we generally define actions, to the point that it is confusing.

I still feel that making it non-optional would be best. I wonder if there is a way that you can solve the problem of `loadVersionInfo` being used frequently in tests with a helper? If not, perhaps the default can be set in the `case` statement for `loadVersionInfo`, rather than in the action definition?

What do you think?

mirefly replied:
>  I find this way of doing it to be inconsistent with the way we generally define actions, to the point that it is confusing.

I agree.

@bobsilverberg I come up with this approach, but not very sure. How do you think about it?
``` diff
diff --git a/src/reducers/versions.spec.tsx b/src/reducers/versions.spec.tsx
index 53d499d..7ad545f 100644
--- a/src/reducers/versions.spec.tsx
+++ b/src/reducers/versions.spec.tsx
@@ -13,7 +13,7 @@ import reducer, {
   Version,
   VersionEntryStatus,
   VersionEntryType,
-  actions,
+  _actions as actions,
   createEntryStatusMap,
   createInternalDiff,
   createInternalHunk,
diff --git a/src/reducers/versions.tsx b/src/reducers/versions.tsx
index 9978794..8c54da2 100644
--- a/src/reducers/versions.tsx
+++ b/src/reducers/versions.tsx
@@ -242,13 +242,10 @@ export const actions = {
       resolve(payload);
   }),
   loadVersionInfo: createAction('LOAD_VERSION_INFO', (resolve) => {
-    return ({
-      updatePathInfo = true,
-      ...moreParams
-    }: {
+    return (payload: {
       version: ExternalVersionWithContent | ExternalVersionWithDiff;
-      updatePathInfo?: boolean;
-    }) => resolve({ updatePathInfo, ...moreParams });
+      updatePathInfo: boolean;
+    }) => resolve(payload);
   }),
   loadEntryStatusMap: createAction('LOAD_ENTRY_STATUS_MAP', (resolve) => {
     return (payload: {
@@ -344,6 +341,17 @@ export const actions = {
   ),
 };
 
+export const _actions: typeof actions = {
+  ...actions,
+  loadVersionInfo: ({
+    updatePathInfo = true,
+    version,
+  }: {
+    updatePathInfo?: boolean;
+    version: ExternalVersionWithContent | ExternalVersionWithDiff;
+  }) => actions.loadVersionInfo({ updatePathInfo, version }),
+};
+
 export type VersionsState = {
   byAddonId: {
     [addonId: number]: VersionsMap;
@@ -1098,6 +1106,7 @@ export const fetchDiff = ({
         dispatch(
           actions.loadVersionInfo({
             version: response,
+            updatePathInfo: true,
           }),
         );
       }

```

bobsilverberg replied:
This is clever, but I fear it may be a bit too clever. It still differs greatly from anything else we have, and that can make it difficult to follow for anyone looking at the code for the first time.

What if you were to create a helper called `_loadVersionInfo` in the tests that calls `actions.loadVersionInfo` with a default of `true` for `updatePathInfo`? That might remove the burden from the tests and still allow us to define `updatePathInfo` as a non-optional argument in a standard way.

mirefly replied:
I am not sure where to define `_loadVersionInfo`, since it will be used in a few test files, `test-helper.tsx` or `version.tsx` (or `version.spec.tsx`). But in either way, we have to import `_loadVersionInfo` as a special case, different from all other actions (I do not like it...). Let me think about it again.

bobsilverberg replied:
Yes, I see your point. Could you post a few links to places where we currently need to use `loadVersionInfo` in tests and I can see if I can come up with an idea of making a helper work that will feel good to both of us?

mirefly replied:
Here are some places we called `loadVersionInfo` in tests.
https://github.com/mozilla/addons-code-manager/blob/master/src/components/FileTree/index.spec.tsx#L36
https://github.com/mozilla/addons-code-manager/blob/master/src/components/Navbar/index.spec.tsx#L317
https://github.com/mozilla/addons-code-manager/blob/master/src/components/VersionFileViewer/index.spec.tsx#L68
https://github.com/mozilla/addons-code-manager/blob/master/src/pages/Browse/index.spec.tsx#L96
https://github.com/mozilla/addons-code-manager/blob/master/src/pages/Compare/index.spec.tsx#L428
https://github.com/mozilla/addons-code-manager/blob/master/src/reducers/fileTree.spec.tsx#L92
https://github.com/mozilla/addons-code-manager/blob/master/src/reducers/versions.spec.tsx#L226
https://github.com/mozilla/addons-code-manager/blob/master/src/test-helpers.tsx#L873

bobsilverberg replied:
Thanks @mirefly. I'll take a look and see if I have any good ideas. In the meantime, what did you think about the idea of dealing with the default value in the `case` statement?

mirefly replied:
> what did you think about the idea of dealing with the default value in the case statement?

That will be easiest way! I am Ok with that approach.
I just want to make sure you mean something like this, right?
```diff

diff --git a/src/reducers/versions.tsx b/src/reducers/versions.tsx
index acf5c2e..11b4957 100644
--- a/src/reducers/versions.tsx
+++ b/src/reducers/versions.tsx
@@ -245,6 +245,7 @@ export const actions = {
   loadVersionInfo: createAction('LOAD_VERSION_INFO', (resolve) => {
     return (payload: {
       version: ExternalVersionWithContent | ExternalVersionWithDiff;
+      updatePathInfo?: boolean;
     }) => resolve(payload);
   }),
   loadEntryStatusMap: createAction('LOAD_ENTRY_STATUS_MAP', (resolve) => {
@@ -1257,6 +1258,19 @@ export const createReducer = ({
       case getType(actions.loadVersionInfo): {
         const { version } = action.payload;
         const internalVersion = createInternalVersion(version);
+
+        let { updatePathInfo } = action.payload;
+        updatePathInfo = updatePathInfo ? updatePathInfo : true;
+        if (updatePathInfo) {
+          return {
+            ...state,
+            versionInfo: {
+              ...state.versionInfo,
+              [version.id]: internalVersion,
+            },
+          };
+        }
+
         const {
           expandedPaths: expandedPathsInState,
           selectedPath: selectedPathInState,


```

mirefly replied:
@bobsilverberg I reconsidered all approaches we discussed so far. I personally like the idea of dealing with default value in `case` statement most.

I might have to add another optional parameter for `loadVersionInfo` in fixing #932 . It is unpleasant to update default values in several places.

mirefly commented:
@bobsilverberg  Another approach might be to add a new action like `loadVersionInfoOnly` which does not update path info, but it will introduce a bit duplication in `loadVersionInfo` and `loadVersionInfoOnly`. How do you think about it?
<details><summary> patch</summary>

```diff
diff --git a/src/reducers/versions.tsx b/src/reducers/versions.tsx
index acf5c2e..2ddd3e0 100644
--- a/src/reducers/versions.tsx
+++ b/src/reducers/versions.tsx
@@ -247,6 +247,11 @@ export const actions = {
       version: ExternalVersionWithContent | ExternalVersionWithDiff;
     }) => resolve(payload);
   }),
+  loadVersionInfoOnly: createAction('LOAD_VERSION_INFO_ONLY', (resolve) => {
+    return (payload: {
+      version: ExternalVersionWithContent | ExternalVersionWithDiff;
+    }) => resolve(payload);
+  }),
   loadEntryStatusMap: createAction('LOAD_ENTRY_STATUS_MAP', (resolve) => {
     return (payload: {
       version: ExternalVersionWithContent | ExternalVersionWithDiff;
@@ -815,7 +820,11 @@ export const fetchVersion = ({
       dispatch(actions.abortFetchVersion({ versionId }));
       dispatch(errorsActions.addError({ error: response.error }));
     } else {
-      dispatch(actions.loadVersionInfo({ version: response }));
+      if (setAsCurrent) {
+        dispatch(actions.loadVersionInfo({ version: response }));
+      } else {
+        dispatch(actions.loadVersionInfoOnly({ version: response }));
+      }
       dispatch(
         actions.loadVersionFile({
           version: response,
@@ -1285,6 +1294,16 @@ export const createReducer = ({
           },
         };
       }
+      case getType(actions.loadVersionInfoOnly): {
+        const { version } = action.payload;
+        return {
+          ...state,
+          versionInfo: {
+            ...state.versionInfo,
+            [version.id]: createInternalVersion(version),
+          },
+        };
+      }
       case getType(actions.abortFetchVersion): {
         const { versionId } = action.payload;
```
</details>

Regarding with creating a `helper`, I am considering something like the following, but I am not sure if it is better or worse than the simply creating `_loadVersionInfo` in `test-helpers.tsx`.
```diff
diff --git a/src/test-helpers.tsx b/src/test-helpers.tsx
index ec0ce83..e783567 100644
--- a/src/test-helpers.tsx
+++ b/src/test-helpers.tsx
@@ -1030,3 +1030,15 @@ export const fakeChangeInfo: ChangeInfo = Object.freeze({
   oldLineNumber: 1,
   type: 'insert',
 });
+
+export const createVersionActions = ({
+  updatePathInfo,
+}: {
+  updatePathInfo: boolean;
+}) => {
+  return {
+    ...versionActions,
+    loadVersionInfo: ({ version, updatePathInfo }) =>
+      versionActions.loadVersionInfo({ version, updatePathInfo }),
+  };
+};

```

bobsilverberg commented:
> @bobsilverberg Another approach might be to add a new action like `loadVersionInfoOnly` which does not update path info, but it will introduce a bit duplication in `loadVersionInfo` and `loadVersionInfoOnly`. How do you think about it?

Interesting idea, but I think I prefer your initial implementation with a param for `updatePathInfo`.

> patch
> 
> Regarding with creating a `helper`, I am considering something like the following, but I am not sure if it is better or worse than the simply creating `_loadVersionInfo` in `test-helpers.tsx`.
> 
> ```diff
> diff --git a/src/test-helpers.tsx b/src/test-helpers.tsx
> index ec0ce83..e783567 100644
> --- a/src/test-helpers.tsx
> +++ b/src/test-helpers.tsx
> @@ -1030,3 +1030,15 @@ export const fakeChangeInfo: ChangeInfo = Object.freeze({
>    oldLineNumber: 1,
>    type: 'insert',
>  });
> +
> +export const createVersionActions = ({
> +  updatePathInfo,
> +}: {
> +  updatePathInfo: boolean;
> +}) => {
> +  return {
> +    ...versionActions,
> +    loadVersionInfo: ({ version, updatePathInfo }) =>
> +      versionActions.loadVersionInfo({ version, updatePathInfo }),
> +  };
> +};
> ```

After looking at some of the links you added above, it seems that a lot of the time we are dispatching `loadVersionInfo` in order to set up data for a test. This suggests to me that we could create a helper called `dispatchVersionInfo`, which would accept an argument for `store` as well as `version` and optionally `updatePathInfo`, and it would simply dispatch `loadVersionInfo` with the appropriate params, allowing `updatePathInfo` to default to `true`. Then, whenever a test needs to load version data for set up purposes, it could call `dispatchVersionInfo` instead.

My thinking is that the test doesn't really need to know that `loadVersionInfo` exists inside `actions` - it can just call the helper as needed. This actually decouples the tests from the implementation of the `versions` reducer, which seems like an additional benefit.

What do you think of that idea?

mirefly commented:
> This suggests to me that we could create a helper called dispatchVersionInfo

Good idea! Let met try it a bit to see if it works.

mirefly commented:
Thanks @bobsilverberg for helping on this. I updated the patch for another look.
It looks like that I need both   `_loadVersionInfo` in `reducers/versions.spec.tsx` and `dispatchLoadVersionInfo`  for other tests (I guess you mean `dispatchLoadVersionInfo` not `dispatchVersionInfo` right?). As to the other two places calling `loadVersionInfo` in `reducers/fileTree.spec.tsx`, I just added the param `updatePathInfo: true`.

I am not fully satisfied, since default value has to be set in at least two places `dispatchLoadVersionInfo` and `_loadVersionInfo`. But I cannot think of a better solution to set default param for `loadVersionInfo` in all tests.

bobsilverberg commented:
Thank you @mirefly for your patience as we try to come up with an ideal solution!

> Thanks @bobsilverberg for helping on this. I updated the patch for another look.
> It looks like that I need both `_loadVersionInfo` in `reducers/versions.spec.tsx` and `dispatchLoadVersionInfo` for other tests (I guess you mean `dispatchLoadVersionInfo` not `dispatchVersionInfo` right?). As to the other two places calling `loadVersionInfo` in `reducers/fileTree.spec.tsx`, I just added the param `updatePathInfo: true`.
> 

Yes, that all sounds good, and I think having a  `_loadVersionInfo` in `reducers/versions.spec.tsx` is just fine.

> I am not fully satisfied, since default value has to be set in at least two places `dispatchLoadVersionInfo` and `_loadVersionInfo`. But I cannot think of a better solution to set default param for `loadVersionInfo` in all tests.

Again, I think that's fine. The two helpers fulfill different requirements (one is to dispatch an action that is needed for the test, and the other is to verify an action).

mirefly commented:
@bobsilverberg  I requested a review in case it is not in your queue yet. No rush.
